DOC: Building libtorch using CMake (#44196)
pytorch/pytorch: docs/libtorch.rst

To use ninja rather than make, set `CMAKE_GENERATOR="-GNinja" CMAKE_INSTALL="ninja install"`.

Note that we are working on eliminating tools/build_pytorch_libs.sh in favor of a unified cmake build.

> Building libtorch without using Python

Reviewer: ATen, which is part of libtorch,  can not be built from source  if Python3 runtime + dataclasses + pyyaml are missing on the system, see https://github.com/pytorch/pytorch/blob/24df3b7373a80fbccebfde72f70fdd045ec44430/cmake/Codegen.cmake#L185
```suggestion
Building libtorch using CMake
```

--------------------------------------

> You can build C++ libtorch.so without relying on Python.  For example to build a v1.6.0 Release version that uses (system-installed) CUDA and CUDDN, and install it in the directory specified by CMAKE_INSTALL_PREFIX below, you can use

Reviewer: Nit
```suggestion
You can build C++ libtorch.so without relying on Python.  For example to build a Release version, with CUDA support(if available), and install it in the directory specified by CMAKE_INSTALL_PREFIX below, you can use
```

Author: Please see the most recent commit.

::

>    git clone https://github.com/pytorch/pytorch.git
>    cd pytorch
>    git checkout v1.6.0
>    git clean -fdx
>    git submodule update --init --recursive

Reviewer: Why one need to call `clean -fdx` on a freshly checked out repo?
Also, in order to avoid changing documentation with every release, it probably better to document how to build a latest master
```suggestion
   git clone --recurse-submodules https://github.com/pytorch/pytorch.git
   cd pytorch
```

Author: The initial `clone` as written will bring in the `master` branch.  The checkout of an earlier tag will make changed files revert to their version at the time of the tag, but won't always delete files and directories that were not present for that tag.  The `git clean -fdx` covers us for that case, so that the subsequent build won't try to use files or directories that shouldn't be there.

There is value in showing how to do this with `master` but many users won't want the bleeding edge implementation, rather a stable tagged version.  We could show both.  I'll put in a commit with that.

Author: Please see the most recent commit.

Reviewer: `git clone -b` can be used to checkout specific tag/branch.
My concerns with referencing specific release tag is that it will require updating the README with every release.
Besides there are already binary package for release tags  available at https://www.pytorch.org, so there isn't much added value in rebuilding them locally, is there?

   mkdir ../pytorch-build
   cd ../pytorch-build

>    cmake -DBUILD_SHARED_LIBS:BOOL=ON -DCMAKE_BUILD_TYPE:STRING=Release -DUSE_CUDA:BOOL=ON -DUSE_CUDNN:BOOL=ON -DCMAKE_INSTALL_PREFIX:PATH=../pytorch-install ../pytorch

Reviewer: No need to specify configuration option that should be auto-detected (CMake will try to build with CUDA/CUDNN by default if it is available on the system search path)
```suggestion
   cmake  -DCMAKE_BUILD_TYPE=Release -DPYTHON_EXECUTABLE=`which python3` -DCMAKE_INSTALL_PREFIX:PATH=../pytorch-install ../pytorch
```

Author: I explicitly listed some options with their default values.  That makes it easier for the user to see how to make changes, to other than default values.  I can add -DPYTHON_EXECUTABLE=`which python3`

Author: Please see the most recent commit.

Reviewer: Can you please elaborate, how do you decide, which options to include and which aren't?
Wouldn't it be easier to read/understand if default instructions would be as concise as possible and then have an additional paragraphs covering useful configurations, for example:
 - How to build  PyTorch with CUDA support if default gcc is too new (by switching to clang or using cuda-gcc from rpmfusion), i.e. by specifing: `-DCMAKE_CXX_COMPILER=/usr/bin/cuda-g++ -DCMAKE_C_COMPILER=/usr/bin/cuda-gcc`
- How to build PyTorch with address sanitizer (-DUSE_ASAN) or in RelWithDebInfo mode, which gives one a performant runtime but at the same time a powerful features such as symbolicated backtrace, line information and even some frame variable information.

To use ninja rather than make, set `CMAKE_GENERATOR="-GNinja" CMAKE_INSTALL="ninja install"`.

Note that we are working on eliminating tools/build_pytorch_libs.sh in favor of a unified cmake build.

Building libtorch using CMake
--------------------------------------

You can build C++ libtorch.so directly with cmake.  For example, to build a v1.6.0 Release version that uses CUDA and CUDDN, and install it in the directory specified by CMAKE_INSTALL_PREFIX below, you can use
::
   git clone https://github.com/pytorch/pytorch.git
   cd pytorch
   git checkout v1.6.0
   git clean -fdx
   git submodule update --init --recursive
   mkdir ../pytorch-build
   cd ../pytorch-build

>    cmake -DBUILD_SHARED_LIBS:BOOL=ON -DCMAKE_BUILD_TYPE:STRING=Release -DPYTHON_EXECUTABLE=`which python3` -DUSE_CUDA:BOOL=ON -DUSE_CUDNN:BOOL=ON -DCMAKE_INSTALL_PREFIX:PATH=../pytorch-install ../pytorch

Reviewer: Is there a reason why you specify variable types in some cases (i.e. `-DBUILD_SHARED_LIBS:BOOL=ON`), but do not in others (i.e. -DPYTHON_EXECUTABLE=foo vs -DPYTHON_EXECUTABLE:PATH=FOO)

Author: I will change that to `-DPYTHON_EXECUTABLE:PATH=FOO`.

   cmake --build . --target install

You can build the ``Debug`` version instead of ``Release`` and you can turn options off with ``OFF`` instead of ``ON``.  Currently USE_CUDNN and USE_CUDA default to ``ON`` when those packages are available, and these and other options need not be specified when the default values are desired.

> Also note that if you want to build the current ``master`` version then the ``git checkout ...`` and ``git clean ...`` commands are not needed.  Furthermore in this case, ``git submodule ...`` is not needed if the ``git clone ...`` command is changed to ``git clone --recurse-submodules https://github.com/pytorch/pytorch.git`` instead.

Author: Will do.
